api/core/tools/provider/builtin/yahoo/tools/news.py

```diff
@@ -0,0 +1,46 @@
+from core.tools.tool.builtin_tool import BuiltinTool
+from core.tools.entities.tool_entities import ToolInvokeMessage
+
+from typing import Any, Dict, List, Union
+from requests.exceptions import HTTPError, ReadTimeout
+
+import yfinance
+
+class YahooFinanceSearchTickerTool(BuiltinTool):
+    def _invoke(self,user_id: str, tool_paramters: Dict[str, Any]) \
+          -> Union[ToolInvokeMessage, List[ToolInvokeMessage]]:
+        '''
+            invoke tools
+        '''
+        
+        query = tool_paramters.get('symbol', '')
+        if not query:
+            return self.create_text_message('Please input symbol')
+        
+        try:
+            return self.run(ticker=query, user_id=user_id)
+        except (HTTPError, ReadTimeout):
+            return self.create_text_message(f'There is a internet connection problem. Please try again later.')
+
+    def run(self, ticker: str, user_id: str) -> ToolInvokeMessage:
+        company = yfinance.Ticker(ticker)
+        try:
+            if company.isin is None:
+                return self.create_text_message(f'Company ticker {ticker} not found.')
+        except (HTTPError, ReadTimeout, ConnectionError):
+            return self.create_text_message(f'Company ticker {ticker} not found.')
+
+        links = []
+        try:
+            links = [n['link'] for n in company.news if n['type'] == 'STORY']
+        except (HTTPError, ReadTimeout, ConnectionError):
+            if not links:
+                return self.create_text_message(f'There is nothing about {ticker} ticker')
+        if not links:
+            return self.create_text_message(f'No news found for company that searched with {ticker} ticker.')
+        
+        result = '\n\n'.join([
+            self.get_url(link) for link in links
+        ])
+
+        return self.create_text_message(self.summary(user_id=user_id, content=result))
```

Comment:
Could it be that this is consuming a lot of tokens when `result` is huge, ignoring the max tokens defined in the dify app model settings?
When I tried this, it took like 10 seconds to complete, and I see a lot of data in the dify logs (6000+ chars). It also looks like a LLM is generating this text and then the yahoo tool is generating a summary on top.

Example from logs (only 2k chars from content):

```
[on_llm_before_invoke]
Model: mistralai/Mixtral-8x7B-Instruct-v0.1
Parameters:
        temperature: 0.8
        top_p: 0.8
Stream: False
User: b8c92f61-6dcb-4b73-973c-4de53c3816f3
Prompt messages:
        role: system
        content: You are a professional language researcher, you are interested in the language
and you can quickly aimed at the main point of an webpage and reproduce it in your own words but 
retain the original meaning and keep the key points. 
however, the text you got is too long, what you got is possible a part of the text.
Please summarize the text you got.

        role: user
        content: Artificial intelligence (AI) burst onto the scene last year, making headlines and sparking public interest. While the technology has been around in some form or another since the 1950s, recent advances have produced models that can create original content and streamline processes, resulting in dramatic productivity gains. This has captured the attention of Main Street and Wall Street alike.Veteran Wedbush tech analyst Dan Ives said generative AI could spark the "fourth industrial revolution," adding "AI is the most transformational technology we have seen since the internet started to take shape."That's a bold view but one that's spreading through tech like wildfire. Even the most conservative estimates suggest that generative AI could generate more than $1 trillion in incremental spending over the coming decade. Companies positioned to profit from this paradigm shift will capture part of the windfall, enriching shareholders along the way.My portfolio is rich with AI stocks, but the one I'm most excited about right now is Microsoft (NASDAQ: MSFT).Image source: Getty Images.The once and future kingEarlier branches of AI have been used extensively in the tech industry for decades. Recommendations made by streaming services and online retailers, voice and facial recognition on smartphones, and maps and navigation software are just a few of a multitude of ways AI is already enriching our lives.The coming of generative AI takes that to the next level. Generative AI can create original text and music, summarize data, draft email responses, create presentations from existing data, and even write and debug computer code. Novel uses for the technology are being developed every day.Microsoft recognized the utility of generative AI early on, investing $1 billion in ChatGPT creator OpenAI as early as 2019, increasing its stake to $13 billion last year. The company quickly integrated this technology into many of its flagship products, imbuing them with AI functionality, resulting in increased productivity for users. Generative AI also has the potential to generate billions of dollars in incremental revenue for Microsoft.
                 [...]
```

This user message is not from me, I assume this is generated by an agent?

Reply:
Summarizing a long text before sending it to LLM is required, otherwise it causes an out of context error, the `summary` method will detect if there is a need to summarize the content automatically based on context size of LLM, have you got any idea of resolving this problem?

In my view, one solution is limiting the `page_size` of yahoo finance, for example, limit at most 3 pages, but it causes some news missing.

Reply:
We could consider adding a setting for whether to use the summary strategy. You could also try using the Truncate END strategy as an alternative. But he results might not be as expected.

Reply:
> Summarizing a long text before sending it to LLM is required, otherwise it causes an out of context error,

Yes, that makes sense. The issue here is, that this can lead to uncontrolled token usage, if the text to summarize is large. And from what I've seen it wasn't really a good summarization since the model still generated 6000+ chars?

> > have you got any idea of resolving this problem?
> you could also try using the Truncate END strategy as an alternative. But he results might not be as expected.
Yes, I'd prefer that but I would add a max length configurable input to the tool so that users can configure that if they wish to set it. We could inform the model, that the text has been truncated and that it's only summarizing e.g. 400 of 1000 chars. Doing this the LLM can still inform the user about that.

In addition to that it would be great if it would be possible to define a System Summarization Model that is beeing used to generate this summarizations. Because I might want to use a cheaper model to summarize. And maybe also allow to configure it per app together with the summarization instruction so that I e.g. can say "summarize the text to have 100-200 chars." which allows me to save tokens in the summarization too.